src/data_member.rs

```diff
@@ -72,6 +72,108 @@ impl<T: Sized + Copy> DataMember<T> {
+    /// of offsets, then you want to put the base address of the module as `addr`, `0x12345678` as
+    /// the first offset, then any further offsets etc.
+    /// This function is merely a convenience function, and is equivalent to
+    /// `new_offset_relative(handle, vec![addr as isize, offsets[0], offsets[1], ...])`.
+    /// You must
+    /// remember to call [`try_into_process_handle`] on a [`Pid`] as sometimes the `Pid` can have
+    /// the same backing type as a [`ProcessHandle`], resulting in an error.
+    ///
+    /// [`try_into_process_handle`]: trait.TryIntoProcessHandle.html#tymethod.try_into_process_handle
+    /// [`ProcessHandle`]: type.ProcessHandle.html
+    /// [`Pid`]: type.Pid.html
+    #[must_use]
+    #[allow(clippy::doc_markdown)]
+    #[allow(clippy::cast_sign_loss)]
+    pub fn new_addr_offset(handle: ProcessHandle, addr: usize, offsets: Vec<isize>) -> Self {
+        let mut vec = vec![addr];
+        // Yes, we are casting to usize. This will not touch any bits, and due to 2s complement,
+        // we still get the correct result when adding offsets.
+        vec.extend(offsets.into_iter().map(|x| x as usize));
+        Self {
+            offsets: vec,
+            process: handle,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+
+    /// Creates a new `DataMember` by appending some more offets. Useful when you have a data
+    /// structure, and want to refer to multiple fields in it, or use it as a starting point
+    /// for chasing down more pointers.
+    /// Since the pointed-to data type might have changed, this function is generic. It is your
+    /// responsibility to make sure you know what you point to.
+    #[allow(clippy::cast_sign_loss)]
+    #[must_use]
+    pub fn extend<TNew>(&self, more_offsets: Vec<isize>) -> DataMember<TNew> {
+        let mut clone = DataMember {
+            offsets: self.offsets.clone(),
+            process: self.process,
+            _phantom: std::marker::PhantomData,
+        };
+        // Yes, we are casting to usize. This will not touch any bits, and due to 2s complement,
+        // we still get the correct result when adding offsets.
+        clone
+            .offsets
+            .extend(more_offsets.into_iter().map(|x| x as usize));
+        clone
+    }
+
+    /// Creates a new `DataMember`, based on self, by shifting the last offset by a number of
+    /// bytes. Does not append new offsets. This is useful if you have a pointer to a struct
+    /// and want to address different fields, or access elements in an array.
+    #[allow(clippy::cast_sign_loss)]
+    #[must_use]
+    pub fn shift<TNew>(&self, n_bytes: isize) -> DataMember<TNew> {
+        let mut clone = DataMember {
+            offsets: self.offsets.clone(),
+            process: self.process,
+            _phantom: std::marker::PhantomData,
+        };
+        let new = clone.offsets[self.offsets.len() - 1].wrapping_add(n_bytes as usize);
```

Comment:
The number of offsets needs to be checked here, or else you'll panic if someone just called `DataMember::new()` and passed it to this function.

I'm not sure of the best way to handle this. Maybe returning an `Option<DataMember<...>>` and doing something like `let new = clone.offsets.last_mut()?.wrapping_add(n_bytes as usize);`?

Reply:
Another potential option would be turning it into `extend()` if there are no offsets in the existing data member.